### PR TITLE
Diff against the old live state to enable drift correction

### PR DESCRIPTION
**Pull request by rquitales**
Previously, external changes to Kubernetes resources managed by Pulumi would not be drift correctly by this provider when running `pulumi up --refresh`. This is because we were diff'ing the new inputs against the old inputs. In this scenario, Pulumi would not detect any changes. This PR addresses this by diffing against the old live state when we run `pulumi up`.

Fixes: #2404

### CODE REVIEW

File: tests/sdk/nodejs/nodejs_test.go [hunk 1]
```diff
@@ -1338,3 +1338,145 @@ func TestServiceAccountTokenSecret(t *testing.T) {
...
+	err := pt.TestLifeCycleInitAndDestroy()
+	assert.NoError(t, err)
+
+	// Sanity check with kubectl to verify that the ConfigMap was created with the wanted label.
+	out, err := exec.Command("kubectl", "get", "configmap", "-o", "yaml", "-n", ns, cmName).CombinedOutput()
+	assert.NoError(t, err)
+	assert.Contains(t, string(out), "bar") // ConfigMap should have been created with data foo: bar.
+
+	// Update the ConfigMap and change the data foo: bar to foo: baz.
+	out, err = exec.Command("kubectl", "patch", "configmap", "-n", ns, cmName, "-p", `{"data":{"foo":"baz"}}`).CombinedOutput()
```
Comment (lblackstone) [inline tests/sdk/nodejs/nodejs_test.go:1391]: I just checked, and `kubectl patch` does not update the `lastAppliedConfiguration` annotation, which is used for CSA diffing. See https://github.com/pulumi/pulumi-kubernetes/issues/694 for a related issue.

**Review by lblackstone — CHANGES_REQUESTED** [review]
I'm wary of this change since it affects old / well-tested code, and alters the core diffing functionality. I haven't had a chance to review in-depth, but some particular things to watch out for:

1. CSA and SSA diffing work differently; CSA diff used the `lastAppliedConfiguration` annotation, while SSA diffing is based on apiserver responses, with a fallback to CSA during preview if the cluster is unavailable.
2. Live state is not always available during previews; any changes need to account for that

That said, this change may indeed fix a longstanding bug, but we need to be extra careful here.

**Comment by rquitales** [comment]
Converting this to a draft PR for now. This will require slightly more work than what is in this first PR attempt.
More context: https://github.com/pulumi/pulumi-kubernetes/issues/2404#issuecomment-1548525935

File: tests/sdk/nodejs/nodejs_test.go (hunk 1 above)
Comment (rquitales) [reply tests/sdk/nodejs/nodejs_test.go:1391]: Only `kubectl apply` attaches that annotation iirc. This bug would also be triggered by out of band resource modifications by other controllers/tools that use client-go, or other libraries.

**Comment by lblackstone** [comment]
Closing in favor of https://github.com/pulumi/pulumi-kubernetes/pull/2445

**Pull request closed without merging**